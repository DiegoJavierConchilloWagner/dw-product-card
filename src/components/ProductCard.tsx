import React, { createContext, CSSProperties } from 'react';
import { useProduct } from '../hooks/useProduct';
import { ProductContextProps, Product, onChangeArgs, InitialValues, ProductcardHandlers } from '../interfaces/product.interfaces';
import styles from '../styles/styles.module.css';

export const ProductContext = createContext({} as ProductContextProps);
const { Provider } = ProductContext;

export interface Props {
    product: Product;
    // children?: ReactElement | ReactElement[] | () => JSX.Element;
    // children?: () => JSX.Element;
    children?: (args: ProductcardHandlers) => JSX.Element;
    className?: string;
    style?: CSSProperties;
    onChange?: ( args:onChangeArgs ) => void;
    incBy?: number;
    initialValues?: InitialValues;
}

export const ProductCard = ({ children, product, className, style, onChange, incBy, initialValues }:Props) => {

    const { counter, increaseOrDecreaseBy, maxCount, isMaxCountReached, reset } 
        = useProduct( { onChange,product, incBy, initialValues } );

    return (
        <Provider value={{ counter, increaseOrDecreaseBy, maxCount, product }}>
            <div className={`${styles.productCard} ${className}`} style={style}>
            {
                children && children({
                    count: counter, 
                    isMaxCountReached,
                    maxCount: initialValues?.maxCount,
                    product,
                    
                    increaseOrDecreaseBy,
                    reset
                })
            }
            </div>
        </Provider>
    )
}

