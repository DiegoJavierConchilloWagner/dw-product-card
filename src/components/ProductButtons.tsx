import React, { useCallback, useContext } from "react";
import { ProductContext } from "./ProductCard";
import styles from '../styles/styles.module.css';

export interface Props {
    className?: string;
    style?: React.CSSProperties;
}

export const ProductButtons = ({ className, style }:Props) => {

    const { increaseOrDecreaseBy, counter, maxCount } = useContext( ProductContext );

    const isMaxReached = useCallback(
        () => !!maxCount && counter === maxCount,
        [counter, maxCount],
    );

    return(
        <div className={`${styles.buttonsContainer} ${className}`} style={style} >
            <button className={styles.buttonMinus} onClick={ () => increaseOrDecreaseBy(-1) }> - </button>
            <div className={styles.countLabel}> {counter} </div>
            <button className={`${styles.buttonAdd} ${ isMaxReached() && styles.disabled}`} 
                disabled={isMaxReached()} onClick={ () => increaseOrDecreaseBy(1) } > + </button>
        </div>
    )
};