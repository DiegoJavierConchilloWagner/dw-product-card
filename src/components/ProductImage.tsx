import React, { useContext } from "react";
import { ProductContext } from "./ProductCard";
import noImage from '../assets/no-image.jpg';
import styles from '../styles/styles.module.css';

export interface Props {
    img?: string;
    className?: string;
    style?: React.CSSProperties;
}

export const ProductImage = ({ img, className, style }:Props) => {

    const { product } = useContext( ProductContext );
    let imgToShow:string;

    img ? imgToShow = img : product.img ?
        imgToShow = product.img : imgToShow = noImage;

    return (
        <img className={`${styles.productImg} ${className}`} src={ imgToShow } alt="" style={style} />
    )
};