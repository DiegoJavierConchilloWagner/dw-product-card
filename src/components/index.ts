import { ProductCard as ProductCardHOC } from './ProductCard';
import { ProductCardhHOCProps } from '../interfaces/product.interfaces';

import { ProductTitle } from '.';
import { ProductImage } from './ProductImage';
import { ProductButtons } from './ProductButtons';

export { ProductButtons } from "./ProductButtons";
export { ProductImage } from "./ProductImage";
export { ProductTitle } from "./ProductTitle";
// export { ProductCard } from './ProductCard';
export const ProductCard: ProductCardhHOCProps = Object.assign( ProductCardHOC, {
    Title: ProductTitle,
    Image: ProductImage,
    Buttons: ProductButtons
});


export default ProductCard;