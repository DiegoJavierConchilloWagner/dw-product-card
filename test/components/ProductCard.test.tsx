import React from 'react';
import renderer from 'react-test-renderer'
import { ProductCard } from '../../src/components';
import { product2 } from '../data/products';

const { act } = renderer;

describe('<ProductImage />', () => {
    test('Debe mostrar el componente correctamente', () => {
        
        const wrapper = renderer.create(
            <ProductCard product={product2}>
                {
                    () => (
                        <h1>Hello World</h1>
                    )
                }
            </ProductCard>
        );

        expect(wrapper.toJSON()).toMatchSnapshot();
    });

    test('Debe incrementar el contador', () => {

        const wrapper = renderer.create(
            <ProductCard product={product2}>
                {
                    ( {count, increaseOrDecreaseBy} ) => (
                        <>
                            <h1>Product Card</h1>
                            <span>{ count }</span>
                            <button onClick={ () => increaseOrDecreaseBy(+1) }>+1</button>
                        </>
                    )
                }
            </ProductCard>
        );

        let tree = wrapper.toJSON();
        expect(tree).toMatchSnapshot();
                
        act( () => {
            (tree as any).children[2].props.onClick();
        })

        tree = wrapper.toJSON();
        expect((tree as any).children[1].children[0]).toBe('1')
    });
});