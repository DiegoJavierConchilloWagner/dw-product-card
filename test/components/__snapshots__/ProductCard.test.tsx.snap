// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductImage /> Debe incrementar el contador 1`] = `
<div
  className="productCard undefined"
>
  <h1>
    Product Card
  </h1>
  <span>
    0
  </span>
  <button
    onClick={[Function]}
  >
    +1
  </button>
</div>
`;

exports[`<ProductImage /> Debe mostrar el componente correctamente 1`] = `
<div
  className="productCard undefined"
>
  <h1>
    Hello World
  </h1>
</div>
`;
